polling_stations/apps/file_uploads/views.py

```diff
@@ -0,0 +1,137 @@
+from marshmallow import ValidationError as MarshmallowValidationError
+
+from councils.models import Council
+
+
+MAX_FILE_SIZE = 150000000  # 150Mb
+S3_BUCKET_NAME = "pollingstations-data-dev"
+
+
+class FileSchema(Schema):
+    name = fields.String(
+        required=True,
+        validate=validate.Regexp(
+            r"(?i)^.+(\.csv|\.tsv)$", error="Unexpected file type"
+        ),
+    )
+    size = fields.Integer(
+        required=True,
+        validate=validate.Range(min=1, max=MAX_FILE_SIZE, error="File too big"),
+    )
+    type = fields.String(
+        required=True,
+        validate=validate.Regexp(r"^text\/.+$", error="Unexpected file type"),
+    )
+
+
+class UploadRequestSchema(Schema):
+    council_id = fields.String(required=True)
+    files = fields.List(
+        fields.Nested(FileSchema()),
+        required=True,
+        validate=validate.Length(min=1, max=2),
+    )
+
+
+def get_s3_client():
+    config = Config()
+    access_key = config.get_value(settings.BOTO_SECTION, "aws_access_key_id")
+    secret_key = config.get_value(settings.BOTO_SECTION, "aws_secret_access_key")
+    return boto3.client(
+        "s3", aws_access_key_id=access_key, aws_secret_access_key=secret_key
+    )
+
+
+class FileUploadView(LoginRequiredMixin, TemplateView):
+    template_name = "file_uploads/index.html"
+
+    @method_decorator(staff_member_required)
+    @method_decorator(ensure_csrf_cookie)
+    def get(self, request, *args, **kwargs):
+        return super().get(self, request, *args, **kwargs)
+
+    def get_context_data(self, **context):
+        context["councils"] = (
+            Council.objects.all()
+            .filter(~Q(council_id__startswith="N09"))
+            .defer("area")
+            .order_by("name")
+        )
+        return context
+
+    def validate_body(self, body):
+        """
+        Do some basic picture checks on request body and files.
+        Obviously these values can be spoofed, so checks on
+        file extension and MIME type etc are just to "fail fast".
+        We will check the files _properly_ in the lambda hook.
+        """
+        try:
+            UploadRequestSchema().load(body)
+        except MarshmallowValidationError:
+            raise DjangoValidationError("Request body did not match schema")
+
+        try:
+            Council.objects.get(pk=body["council_id"])
+        except Council.DoesNotExist as e:
+            raise DjangoValidationError(str(e))
+
+        files = body["files"]
+        if len(files) == 2 and files[0]["name"] == files[1]["name"]:
+            raise DjangoValidationError("Files must have different names")
+
+    @method_decorator(staff_member_required)
+    @method_decorator(csrf_protect)
+    def post(self, request, *args, **kwargs):
+        body = json.loads(request.body)
+
+        try:
+            self.validate_body(body)
+        except DjangoValidationError as e:
+            return JsonResponse({"error": e.message}, status=400)
```

Comment:
I know this isn't the file content, so should be small in normal usage, but is it worth us validating the upload size here, before parsing it as JSON? Either in nginx or in this view, or both?

Reply:
There's already a hard limit on the POST body size in the NGINX cofig: https://github.com/DemocracyClub/polling_deploy/blob/c819ad898ddeae06ae251eb8204a3e0adddd9e53/files/conf/nginx.conf#L61
I reckon that probably covers it tbh.